File: Serene/Serene.Script/Common/Helper/GridEditorDialog.cs

namespace Serene.Common
{
    using jQueryApi;
    using Serenity;
    using System;
    using System.Collections.Generic;
    using System.Runtime.CompilerServices;
    using System.Linq;

    [IdProperty("__id")]
    public abstract class GridEditorDialog<TEntity> : EntityDialog<TEntity>
        where TEntity: class, new()
    {
        public override void Destroy()
        {
            OnSave = null;
            OnDelete = null;
            base.Destroy();
        }

        protected override void UpdateInterface()
        {
            base.UpdateInterface();

            // apply changes button doesn't work properly with in-memory grids yet
            if (applyChangesButton != null)
                applyChangesButton.Hide();
        }

        protected override void SaveHandler(ServiceCallOptions options, Action<ServiceResponse> callback)
        {
            if (OnSave != null)
                OnSave(options, callback);
        }

        protected override void DeleteHandler(ServiceCallOptions<DeleteResponse> options, Action<DeleteResponse> callback)
        {
            if (OnDelete != null)
                OnDelete(options, callback);
        }

        public Action<ServiceCallOptions, Action<ServiceResponse>> OnSave { get; set; }
        public Action<ServiceCallOptions<DeleteResponse>, Action<DeleteResponse>> OnDelete { get; set; }
    }
}

File: Serene/Serene.Web/Modules/Common/Helpers/GridEditorDialog.ts
namespace Serene.Common {

    @Serenity.Decorators.registerClass()
    export class GridEditorDialog<TEntity> extends Serenity.EntityDialog<TEntity, any> {
        protected getIdProperty() { return "__id"; }

        public onSave: (options: Serenity.ServiceOptions<Serenity.SaveResponse>,
            callback: (response: Serenity.SaveResponse) => void) => void;

        public onDelete: (options: Serenity.ServiceOptions<Serenity.DeleteResponse>,
            callback: (response: Serenity.DeleteResponse) => void) => void;

        protected destroy() {
            this.onSave = null;
            this.onDelete = null;
            super.destroy();
        }

        protected updateInterface() {
            super.updateInterface();

            // apply changes button doesn't work properly with in-memory grids yet
            if (this.applyChangesButton) {
                this.applyChangesButton.hide();
            }
        }

        protected saveHandler(options: Serenity.ServiceOptions<Serenity.SaveResponse>,
            callback: (response: Serenity.SaveResponse) => void): void {
            this.onSave && this.onSave(options, callback);
        }

        protected deleteHandler(options: Serenity.ServiceOptions<Serenity.DeleteResponse>,
            callback: (response: Serenity.DeleteResponse) => void): void {
            this.onDelete && this.onDelete(options, callback);
        }
    }
}